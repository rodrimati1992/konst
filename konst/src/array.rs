//! Const equivalents of array functions.

macro_rules! leak_warning {
    () => {
        concat!(
            "# Warning\n",
            "\n",
            "This macro leaks the initialized part of the array\n",
            "if the closure passed to this macro panics or returns early.\n",
            "\n",
            "note: this warning is not relevant if the elements don't need dropping",
            "(e.g: by implementing `Copy`).\n"
        )
    };
}

/// Const equivalent of
/// [`array::map`](https://doc.rust-lang.org/std/primitive.array.html#method.map).
///
#[doc = leak_warning!()]
///
/// # Limitations
///
/// This macro supports mapping from non-Copy arrays if any of these
/// conditions are met about the parameter of the passed-in closure:
/// 1. it's a pattern that only copies Copy fields of each array element
/// 2. it's a `ref` pattern
///
/// [examples of both of the above conditions below](#map-noncopy-example)
///
/// # Example
///
/// ### Basic
///
/// ```rust
/// use konst::array;
///
/// const TRIMMED: [&str; 3] = array::map!(["  foo", "bar  ", "  baz  "], konst::string::trim);
/// assert_eq!(TRIMMED, ["foo", "bar", "baz"]);
///
/// const LENGTHS: [usize; 3] = array::map!(["foo", "hello", "bar baz"], |s| s.len());
/// assert_eq!(LENGTHS, [3, 5, 7]);
///
/// const SQUARED: [u32; 6] = array::map!([1, 2, 3, 4, 5, 6], |x: u32| x.pow(2));
/// assert_eq!(SQUARED, [1, 4, 9, 16, 25, 36]);
///
/// {
///     let input = [3, 5, 8];
///     let output = array::map!(input, |x| -> u64 { x + 2 });
///     assert_eq!(output, [5, 7, 10]);
/// }
///
/// ```
///
/// <span id="map-noncopy-example"> </span>
/// ### Map from non-Copy array
///
/// Demonstrates both ways to map from a non-Copy array.
///
/// ```rust
/// use konst::array;
///
/// struct NonCopy(u32, u32);
///
/// const PRIME_SUMS: [u32; 3] = {
///     let input = [NonCopy(2, 3), NonCopy(5, 7), NonCopy(11, 13)];
///     
///     // demonstrates the first way to map from non-Copy elements
///     array::map!(input, |NonCopy(l, r)| l + r)
/// };
/// assert_eq!(PRIME_SUMS, [5, 12, 24]);
///
/// const FIBB_SUMS: [u32; 3] = {
///     let input = [NonCopy(2, 3), NonCopy(5, 8), NonCopy(13, 21)];
///     
///     // demonstrates the second way to map from non-Copy elements
///     array::map!(input, |ref nc| nc.0 + nc.1)
/// };
/// assert_eq!(FIBB_SUMS, [5, 13, 34]);
///
/// ```
///
pub use konst_kernel::array_map as map;

/// Const equivalent of [`array::from_fn`](core::array::from_fn).
///
#[doc = leak_warning!()]
///
/// # Limitations
///
/// When the array type is annotated, the array type must be one of:
/// - Square brackets (e.g: `from_fn!([usize; 10] => |i| i)`)
/// - A parenthesized type (e.g: `from_fn!((foo::ArrayAlias) => |i| i * 2)`)
/// - A single identifier (e.g: `from_fn!(ArrayAlias => |i| func(i))`)
///
/// # Example
///
/// ```rust
/// use konst::array;
///
/// {
///     const POWERS: [u64; 5] = array::from_fn!(|i| 2u64.pow(i as u32));
///
///     assert_eq!(POWERS, [1, 2, 4, 8, 16]);
/// }
///
/// // Annotating the array type
/// assert_eq!(
///     array::from_fn!([&str; 6] => |i| konst::string::str_up_to("hello", i)),
///     ["", "h", "he", "hel", "hell", "hello"],
/// );
/// ```
///
pub use konst_kernel::array_from_fn as from_fn;
