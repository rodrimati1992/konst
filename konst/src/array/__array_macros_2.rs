
#[doc(hidden)]
#[macro_export]
macro_rules! __array_map_by_val {
    ($array:expr, $($closure:tt)* ) => (
        $crate::__::__parse_closure_1!{
            ($crate::__array_map2__with_parsed_closure)
            ($array,)
            (array_map),
            $($closure)*
        }
    );
}

#[doc(hidden)]
#[macro_export]
macro_rules! __array_map2__with_parsed_closure {
    (
        $array:expr,
        ($($pattern:tt)*) $(-> $ret:ty)? $mapper:block $(,)?
    ) => (match $crate::array::ArrayConsumer::new($array) {
        mut consumer => {

            let mut builder = $crate::array::ArrayBuilder::new();

            builder.infer_length_from_consumer(&consumer);

            while let Some(elem) = consumer.next() {
                let elem = $crate::__::ManuallyDrop::into_inner(elem);
                let $($pattern)* = elem;
                let mapped $(: $ret)? = $mapper;
                builder.push(mapped);
            }
            $crate::__::mem::forget(consumer);

            builder.build()
        }
    })
}

#[doc(hidden)]
#[macro_export]
macro_rules! __array_from_fn2 {
    ($($args:tt)*) => ({
        $crate::__::__split_array_type_and_closure!{
            (($crate::__array_from_fn2__splitted_type_and_closure) ())
            ()
            ($($args)*)
        }
    });
}

#[doc(hidden)]
#[macro_export]
macro_rules! __array_from_fn2__splitted_type_and_closure {
    ($type:tt $($closure_unparsed:tt)*) => {
        $crate::__::__parse_closure_1!{
            ($crate::__array_from_fn_with_parsed_closure)
            ($type)
            (from_fn_),

            $($closure_unparsed)*
        }
    }
}

#[doc(hidden)]
#[macro_export]
macro_rules! __array_from_fn_with_parsed_closure {
    (
        ($($($type:tt)+)?)

        ($($pattern:tt)*) $(-> $ret:ty)? $mapper:block $(,)?
    ) => ({
        let mut i = 0usize;

        let arr $(: $crate::__::__unparenthesize_ty!($($type)*))? =
            $crate::__array_map2__with_parsed_closure!{
                $crate::__::unit_array(),
                (()) $(-> $ret)? {
                    let $($pattern)* = i;
                    i+=1;
                    $mapper
                }
            };

        arr
    });
}

