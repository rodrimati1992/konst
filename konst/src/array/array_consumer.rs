use core::fmt::{self, Debug};
use core::mem::{ManuallyDrop, MaybeUninit};

/// Const analog of [`core::array::IntoIter`]
/// 
/// This isn't called `IntoIter` because it does not implement the [`ConstIntoIter`] trait,
/// as this type does not have [the API that that trait requires](
/// https://docs.rs/konst/latest/konst/iter/trait.ConstIntoIter.html#isiteratorkind )
/// 
/// # Example
/// 
/// ```rust
/// use konst::array::{ArrayBuilder, ArrayConsumer};
/// 
/// use core::mem::ManuallyDrop as MD;
/// 
/// assert_eq!(ARR, [21, 13, 8, 5, 3]);
/// 
/// const ARR: [u32; 5] = reverse([3, 5, 8, 13, 21]);
/// 
/// const fn reverse<T, const LEN: usize>(arr: [T; LEN]) -> [T; LEN] {
///     let mut iter = ArrayConsumer::new(arr);
///     let mut builder = ArrayBuilder::new();
/// 
///     while let Some(item) = iter.next_back() {
///         builder.push(MD::into_inner(item));
///     }
/// 
///     // necessary to avoid "destructor cannot be evaluated at compile-time" error
///     iter.assert_is_empty();
/// 
///     builder.build()
/// }
/// ```
/// 
/// [`ConstIntoIter`]: crate::iter::ConstIntoIter
#[repr(C)]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "rust_1_83")))]
pub struct ArrayConsumer<T, const N: usize> {
    array: [MaybeUninit<T>; N],
    taken_front: usize,
    taken_back: usize,
}

impl<T, const N: usize> ArrayConsumer<T, N> {
    /// Constructs an ArrayConsumer from an array.
    pub const fn new(array: [T; N]) -> Self {
        Self {
            array: array_into_md(array),
            taken_front: 0,
            taken_back: 0,
        }
    }

    /// Constructs an already-consumed ArrayConsumer.
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use konst::array::ArrayConsumer;
    /// 
    /// let mut iter = ArrayConsumer::<u8, 4>::empty();
    /// 
    /// assert_eq!(iter.next(), None);
    /// 
    /// ```
    pub const fn empty() -> Self {
        Self {
            array: crate::maybe_uninit::uninit_array(),
            taken_front: N,
            taken_back: 0,
        }
    }

    const fn is_empty(&self) -> bool {
        (N - self.taken_front - self.taken_back) == 0
    }

    const fn slice_len(&self) -> usize {
        N - self.taken_front - self.taken_back
    }

    /// Asserts that the ArrayConsumer is empty,
    /// allows using ArrayConsumer in const.
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use konst::array::ArrayConsumer;
    /// 
    /// use core::mem::ManuallyDrop as MD;
    /// 
    /// assert_eq!(SUM, 16);
    /// 
    /// const SUM: u64 = summer(ArrayConsumer::new([3, 5, 8]));
    /// 
    /// const fn summer<const N: usize>(mut iter: ArrayConsumer<u64, N>) -> u64 {
    ///     let mut sum = 0u64;
    ///     while let Some(item) = iter.next() {
    ///         sum += MD::into_inner(item);
    ///     }
    ///     iter.assert_is_empty();
    ///     sum
    /// }
    /// ```
    #[track_caller]
    pub const fn assert_is_empty(self) {
        assert!(self.is_empty());

        core::mem::forget(self);
    }

    /// Gets the remainder of the array as a slice
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use konst::array::ArrayConsumer;
    /// 
    /// let mut iter = ArrayConsumer::new([3, 5, 8]);
    /// 
    /// assert_eq!(iter.as_slice(), &[3, 5, 8][..]);
    /// 
    /// assert!(iter.next().is_some());
    /// assert_eq!(iter.as_slice(), &[5, 8][..]);
    /// 
    /// assert!(iter.next().is_some());
    /// assert_eq!(iter.as_slice(), &[8][..]);
    /// 
    /// assert!(iter.next().is_some());
    /// assert_eq!(iter.as_slice(), &[][..]);
    /// 
    /// assert_eq!(iter.next(), None);
    /// assert_eq!(iter.as_slice(), &[][..]);
    /// ```
    pub const fn as_slice(&self) -> &[T] {
        // SAFETY: self.array is guaranteed initialized starting from self.taken_front
        //         up to `N - self.taken_back`
        unsafe {
            let ptr = self.array.as_ptr().add(self.taken_front).cast::<T>();
            core::slice::from_raw_parts(ptr, self.slice_len())
        }
    }

    /// Gets the remainder of the array as a mutable slice
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use konst::array::ArrayConsumer;
    /// 
    /// let mut iter = ArrayConsumer::new([3, 5, 8]);
    /// 
    /// assert_eq!(iter.as_mut_slice(), &mut [3, 5, 8][..]);
    /// 
    /// assert!(iter.next().is_some());
    /// assert_eq!(iter.as_mut_slice(), &mut [5, 8][..]);
    /// 
    /// assert!(iter.next().is_some());
    /// assert_eq!(iter.as_mut_slice(), &mut [8][..]);
    /// 
    /// assert!(iter.next().is_some());
    /// assert_eq!(iter.as_mut_slice(), &mut [][..]);
    /// 
    /// assert_eq!(iter.next(), None);
    /// assert_eq!(iter.as_mut_slice(), &mut [][..]);
    /// ```
    pub const fn as_mut_slice(&mut self) -> &mut [T] {
        // SAFETY: self.array is guaranteed initialized starting from self.taken_front
        //         up to `N - self.taken_back`
        unsafe {
            let ptr = self.array.as_mut_ptr().add(self.taken_front).cast::<T>();
            core::slice::from_raw_parts_mut(ptr, self.slice_len())
        }
    }

    /// Gets a bitwise copy of this ArrayConsumer, requires `T: Copy`.
    pub const fn copy(&self) -> Self 
    where
        T: Copy
    {
        Self {..*self}
    }

    /// Gets the next element from the array
    /// 
    /// Due to limitations of const eval as of Rust 1.83.0,
    /// this function returns a `ManuallyDrop<T>` to be able to return a `T: Drop` in an `Option`,
    /// you'll need to call [`ManuallyDrop::into_inner`] to get `T` and avoid leaking it. 
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use konst::array::ArrayConsumer;
    /// 
    /// use std::mem::ManuallyDrop as MD;
    /// 
    /// let mut iter = ArrayConsumer::new([3, 5, 8]);
    /// 
    /// assert_eq!(iter.as_slice(), &[3, 5, 8][..]);
    /// 
    /// assert_eq!(iter.next(), Some(MD::new(3)));
    /// assert_eq!(iter.as_slice(), &[5, 8][..]);
    /// 
    /// assert_eq!(iter.next(), Some(MD::new(5)));
    /// assert_eq!(iter.as_slice(), &[8][..]);
    /// 
    /// assert_eq!(iter.next(), Some(MD::new(8)));
    /// assert_eq!(iter.as_slice(), &[][..]);
    /// 
    /// assert_eq!(iter.next(), None);
    /// assert_eq!(iter.as_slice(), &[][..]);
    /// ```
    pub const fn next(&mut self) -> Option<ManuallyDrop<T>> {
        if self.is_empty() {
            return None;
        }

        // SAFETY: self.array[self.taken_front] is guaranteed initialized
        let ret = unsafe { self.array[self.taken_front].assume_init_read() };

        self.taken_front += 1;

        Some(ManuallyDrop::new(ret))
    }

    /// Gets the next element from the end of the array
    /// 
    /// Due to limitations of const eval as of Rust 1.83.0,
    /// this function returns a `ManuallyDrop<T>` to be able to return a `T: Drop` in an `Option`,
    /// you'll need to call [`ManuallyDrop::into_inner`] to get `T` and avoid leaking it. 
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use konst::array::ArrayConsumer;
    /// 
    /// use std::mem::ManuallyDrop as MD;
    /// 
    /// let mut iter = ArrayConsumer::new([3, 5, 8]);
    /// 
    /// assert_eq!(iter.as_slice(), &[3, 5, 8][..]);
    /// 
    /// assert_eq!(iter.next_back(), Some(MD::new(8)));
    /// assert_eq!(iter.as_slice(), &[3, 5][..]);
    /// 
    /// assert_eq!(iter.next_back(), Some(MD::new(5)));
    /// assert_eq!(iter.as_slice(), &[3][..]);
    /// 
    /// assert_eq!(iter.next_back(), Some(MD::new(3)));
    /// assert_eq!(iter.as_slice(), &[][..]);
    /// 
    /// assert_eq!(iter.next_back(), None);
    /// assert_eq!(iter.as_slice(), &[][..]);
    /// ```
    pub const fn next_back(&mut self) -> Option<ManuallyDrop<T>> {
        if self.is_empty() {
            return None;
        }

        let index = N - self.taken_back - 1;

        // SAFETY: self.array[index] is guaranteed initialized
        let ret = unsafe { self.array[index].assume_init_read() };
        
        self.taken_back += 1;

        Some(ManuallyDrop::new(ret))
    }
}

impl<T: Debug, const N: usize> Debug for ArrayConsumer<T, N> {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        Debug::fmt(self.as_slice(), fmt)
    }
}

impl<T: Clone, const N: usize> Clone for ArrayConsumer<T, N> {
    fn clone(&self) -> Self {
        let mut this = Self {
            array: crate::maybe_uninit::uninit_array(),
            taken_front: 0,
            taken_back: N,
        };

        for (i, elem) in self.as_slice().iter().cloned().enumerate() {
            this.array[i] = MaybeUninit::new(elem);
            this.taken_back -= 1;
        }

        this
    }
}

impl<T, const N: usize> Drop for ArrayConsumer<T, N> {
    fn drop(&mut self) {
        unsafe {
            let slice_len = self.slice_len();

            let ptr = self.array.as_mut_ptr().cast::<T>();

            core::ptr::slice_from_raw_parts_mut(ptr.add(self.taken_front), slice_len).drop_in_place();
        }
    }
}


#[doc(hidden)]
const fn array_into_md<T, const N: usize>(arr: [T; N]) -> [MaybeUninit<T>; N] {
    unsafe {
        crate::__::__priv_transmute! {[T; N], [MaybeUninit<T>; N], arr}
    }
}
