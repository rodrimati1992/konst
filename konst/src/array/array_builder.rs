use core::fmt::{self, Debug};
use core::mem::{ManuallyDrop, MaybeUninit};

use crate::array::ArrayConsumer;


/// For constructing an array element by element.
/// 
/// # Example
/// 
/// ```rust
/// use konst::array::ArrayBuilder;
/// 
/// assert_eq!(ARR, [1, 1, 2, 3, 5, 8, 13, 21, 34, 55]);
/// 
/// const ARR: [u32; 10] = {
///     let mut builder = ArrayBuilder::new();
///     builder.push(1);
///     builder.push(1);
/// 
///     while !builder.is_full() {
///         let [.., a, b] = *builder.as_slice() else { unreachable!() };
/// 
///         builder.push(a + b);
///     }
/// 
///     builder.build()
/// };
/// ```
#[repr(C)]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "rust_1_83")))]
pub struct ArrayBuilder<T, const N: usize> {
    array: [MaybeUninit<T>; N],
    inited: usize,
}

impl<T, const N: usize> ArrayBuilder<T, N> {
    /// Constructs an empty ArrayBuilder
    pub const fn new() -> Self {
        ArrayBuilder {
            array: crate::maybe_uninit::uninit_array(),
            inited: 0,
        }
    }

    /// The amount of initialized elements in the array
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use konst::array::ArrayBuilder;
    /// 
    /// let mut builder = ArrayBuilder::<_, 3>::new();
    /// 
    /// assert_eq!(builder.len(), 0);
    /// 
    /// builder.push(3);
    /// assert_eq!(builder.len(), 1);
    /// 
    /// builder.push(5);
    /// assert_eq!(builder.len(), 2);
    /// 
    /// builder.push(8);
    /// assert_eq!(builder.len(), 3);
    /// ```
    /// 
    pub const fn len(&self) -> usize {
        self.inited
    }

    /// Whether the array has been fully initialized
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use konst::array::ArrayBuilder;
    /// 
    /// let mut builder = ArrayBuilder::<_, 3>::new();
    /// 
    /// assert!(!builder.is_full());
    /// 
    /// builder.push(3);
    /// assert!(!builder.is_full());
    /// 
    /// builder.push(5);
    /// assert!(!builder.is_full());
    /// 
    /// builder.push(8);
    /// assert!(builder.is_full());
    /// ```
    /// 
    pub const fn is_full(&self) -> bool {
        self.inited == N
    }

    /// Gets the initialized part of the array as a slice
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use konst::array::ArrayBuilder;
    /// 
    /// let mut builder = ArrayBuilder::<_, 3>::new();
    /// 
    /// assert_eq!(builder.as_slice(), [].as_slice());
    /// 
    /// builder.push(3);
    /// assert_eq!(builder.as_slice(), [3].as_slice());
    /// 
    /// builder.push(5);
    /// assert_eq!(builder.as_slice(), [3, 5].as_slice());
    /// 
    /// builder.push(8);
    /// assert_eq!(builder.as_slice(), [3, 5, 8].as_slice());
    /// ```
    /// 
    pub const fn as_slice(&self) -> &[T] {
        // SAFETY: self.array is guaranteed initialized up to self.inited - 1 inclusive
        unsafe {
            core::slice::from_raw_parts(self.array.as_ptr().cast::<T>(), self.inited)
        }
    }

    /// Gets the initialized part of the array as a mutable slice
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use konst::array::ArrayBuilder;
    /// 
    /// let mut builder = ArrayBuilder::<_, 3>::new();
    /// 
    /// assert_eq!(builder.as_mut_slice(), [].as_mut_slice());
    /// 
    /// builder.push(3);
    /// assert_eq!(builder.as_mut_slice(), [3].as_mut_slice());
    /// 
    /// builder.push(5);
    /// assert_eq!(builder.as_mut_slice(), [3, 5].as_mut_slice());
    /// 
    /// builder.push(8);
    /// assert_eq!(builder.as_mut_slice(), [3, 5, 8].as_mut_slice());
    /// ```
    /// 
    pub const fn as_mut_slice(&mut self) -> &mut [T] {
        // SAFETY: self.array is guaranteed initialized up to self.inited - 1 inclusive
        unsafe {
            core::slice::from_raw_parts_mut(self.array.as_mut_ptr().cast::<T>(), self.inited)
        }
    }

    /// Appends `val` to the array.
    /// 
    /// # Panic
    /// 
    /// Panics if `self.len() == N`, i.e.: the array is fully initialized.
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use konst::array::ArrayBuilder;
    /// 
    /// let mut builder = ArrayBuilder::<_, 3>::new();
    /// 
    /// builder.push(3);
    /// builder.push(5);
    /// builder.push(8);
    /// 
    /// assert_eq!(builder.build(), [3, 5, 8]);
    /// ```
    /// 
    pub const fn push(&mut self, val: T) {
        assert!(self.inited < N, "trying to add an element to full array");

        self.array[self.inited] = MaybeUninit::new(val);

        self.inited += 1;
    }

    /// Unwraps this ArrayBuilder into an array.
    /// 
    /// # Panic
    /// 
    /// Panics if `self.len() != N`, i.e.: the array is not fully initialized.
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use konst::array::ArrayBuilder;
    /// 
    /// assert_eq!(ARR, [3, 5, 8]);
    /// 
    /// const ARR: [u8; 3] = {
    ///     let mut builder = ArrayBuilder::new();
    ///     
    ///     builder.push(3);
    ///     builder.push(5);
    ///     builder.push(8);
    /// 
    ///     builder.build()
    /// };
    /// ```
    /// 
    pub const fn build(self) -> [T; N] {
        assert!(self.is_full(), "trying to unwrap a non-fully-initialized array");

        // SAFETY: self.array is guaranteed fully initialized by the fact that
        //         each element is inited in lockstep with incrementing self.inited by 1,
        //         and the assertion above.
        unsafe {
            let mut this = ManuallyDrop::new(self);

            // this cast is guaranteed correct becaue this struct is `#[repr(C))]`
            // and the first field is a `[MaybeUninit<T>; N]`
            (&raw mut this).cast::<[T; N]>().read()
        }
    }

    /// Gets a bitwise copy of this Builder, requires `T: Copy`.
    pub const fn copy(&self) -> Self 
    where
        T: Copy
    {
        Self {..*self}
    }

    /// Helper for inferring the length of the built array from an [`ArrayConsumer`].
    pub const fn infer_length_from_consumer<U>(&self, _consumer: &ArrayConsumer<U, N>) {}
}

impl<T: Debug, const N: usize> Debug for ArrayBuilder<T, N> {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt.debug_struct("ArrayBuilder")
            .field("array", &self.as_slice())
            .field("uninit_len", &(N - self.inited))
            .finish()
    }
}

impl<T: Clone, const N: usize> Clone for ArrayBuilder<T, N> {
    fn clone(&self) -> Self {
        let mut this = Self::new();
        for elem in self.as_slice() {
            this.push(elem.clone());
        }
        this
    }
}

impl<T, const N: usize> Drop for ArrayBuilder<T, N> {
    fn drop(&mut self) {
        unsafe {
            let inited = self.inited;

            let ptr = self.array.as_mut_ptr().cast::<T>();

            core::ptr::slice_from_raw_parts_mut(ptr, inited).drop_in_place();
        }
    }
}
