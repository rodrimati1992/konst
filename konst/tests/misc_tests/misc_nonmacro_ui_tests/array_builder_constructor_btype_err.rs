use konst::array::ArrayBuilder;


fn main() {
    let _ = ArrayBuilder::<String, 3, _>::of_drop().into_copy();
}

