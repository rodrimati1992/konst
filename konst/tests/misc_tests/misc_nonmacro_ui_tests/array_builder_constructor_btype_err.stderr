error[E0277]: the trait bound `String: Copy` is not satisfied
 --> tests/misc_tests/misc_nonmacro_ui_tests/array_builder_constructor_btype_err.rs:5:53
  |
5 |     let _ = ArrayBuilder::<String, 3, _>::of_drop().into_copy();
  |                                                     ^^^^^^^^^ the trait `Copy` is not implemented for `String`
  |
note: required by a bound in `ArrayBuilder::<T, N, D>::into_copy`
 --> src/array/array_builder.rs
  |
  |     pub const fn into_copy(self) -> ArrayBuilder<T, N, NonDrop>
  |                  --------- required by a bound in this associated function
  |     where
  |         T: Copy
  |            ^^^^ required by this bound in `ArrayBuilder::<T, N, D>::into_copy`
